src/main/java/com/terraformersmc/terrestria/surface/CanyonSurfaceBuilder.java

```diff
@@ -0,0 +1,158 @@
+package com.terraformersmc.terrestria.surface;
+
+import com.mojang.datafixers.Dynamic;
+import com.terraformersmc.terraform.noise.OpenSimplexNoise;
+import net.minecraft.block.BlockState;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.Direction;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.chunk.Chunk;
+import net.minecraft.world.gen.surfacebuilder.SurfaceBuilder;
+import net.minecraft.world.gen.surfacebuilder.TernarySurfaceConfig;
+
+import java.util.Random;
+import java.util.function.Function;
+
+public class CanyonSurfaceBuilder extends SurfaceBuilder<CanyonSurfaceConfig> {
+
+	private static final OpenSimplexNoise CLIFF_NOISE = new OpenSimplexNoise(346987);
+	private int seaLevel;
+	private SurfaceBuilder<TernarySurfaceConfig> parent;
+
+	public CanyonSurfaceBuilder(Function<Dynamic<?>, ? extends CanyonSurfaceConfig> function, int seaLevel, SurfaceBuilder<TernarySurfaceConfig> parent) {
+		super(function);
+
+		this.seaLevel = seaLevel;
+		this.parent = parent;
+	}
+
+	private static int underNoiseToLayers(double noise, int cliffLayers) {
+		int underLayers = (int) (noise * 3 + 0.5);
+
+		if (cliffLayers > 5) {
+			underLayers += 1;
+		}
+
+		if (cliffLayers > 10) {
+			underLayers += 3;
+		}
+
+		if (cliffLayers > 15) {
+			underLayers += 4;
+		}
+
+		return underLayers;
+	}
+
+	/**
+	 * "terraces" the input noise (from 0.0 to 1.0) returning an integer from 1 to 40, inclusive
+	 * Domain: [-1.0, 1.0]
+	 * Range: [1, 40]
+	 */
+	private static int cliffNoiseToLayers(double noise) {
+		// Domain transformation:
+		// [0.0, 1.0] -> [0, 60]
+
+		noise = MathHelper.clamp(noise, 0, 1);
+		noise *= 60;
+
+		int height = 1;
+
+		if (noise > 3) {
+			height += 2;
+		}
+
+		if (noise > 5) {
+			height += 3;
+		}
+
+		if (noise > 9) {
+			height += 4;
+		}
+
+		if (noise > 14) {
+			height += 4;
+		}
+
+		if (noise > 20) {
+			height += 5;
+		}
+
+		if (noise > 27) {
+			height += 6;
+		}
+
+		if (noise > 39) {
+			height += 7;
+		}
+
+		if (noise > 55) {
+			height += 8;
+		}
+
+		return height;
+	}
+
+	@Override
+	public void generate(Random rand, Chunk chunk, Biome biome, int x, int z, int vHeight, double noise, BlockState stone, BlockState water, int var11, long seed, CanyonSurfaceConfig config) {
+		if (vHeight < seaLevel + 5) {
+			// In the future make this dig down instead
```

Comment:
hmm, is this a TODO?

Reply:
You wrote this comment so I'm not sure what you wanted to do there. It could be, possibly.

Reply:
Let's go with no